src/cultionet/callbacks.py

```diff
@@ -0,0 +1,163 @@
+        reference_image: Path,
+        out_path: Path,
+        num_classes: int,
+        ref_res: float,
+        resampling,
+        compression: str,
+        write_interval: str = 'batch'
+    ):
+        super().__init__(write_interval)
+
+        self.reference_image = reference_image
+        self.out_path = out_path
+        self.out_path.parent.mkdir(parents=True, exist_ok=True)
+
+        with gw.config.update(ref_res=ref_res):
+            with gw.open(reference_image, resampling=resampling) as src:
+                profile = {
+                    'crs': src.crs,
+                    'transform': src.gw.transform,
+                    'height': src.gw.nrows,
+                    'width': src.gw.ncols,
+                    # distance (+1) + edge (+1) + crop (+1) crop types (+N)
+                    # `num_classes` includes background
+                    'count': 3 + num_classes - 1,
+                    'dtype': 'uint16',
+                    'blockxsize': 64 if 64 < src.gw.ncols else src.gw.ncols,
+                    'blockysize': 64 if 64 < src.gw.nrows else src.gw.nrows,
+                    'driver': 'GTiff',
+                    'sharing': False,
+                    'compress': compression
+                }
+        profile['tiled'] = True if max(profile['blockxsize'], profile['blockysize']) >= 16 else False
+        with rio.open(out_path, mode='w', **profile):
+            pass
+        self.dst = rio.open(out_path, mode='r+')
+
+    def write_on_epoch_end(self, trainer, pl_module, predictions, batch_indices):
+        self.dst.close()
+
+    def reshape_predictions(
+        self,
+        batch: Data,
+        distance_batch: torch.Tensor,
+        edge_batch: torch.Tensor,
+        crop_batch: torch.Tensor,
+        crop_type_batch: T.Union[torch.Tensor, None],
+        batch_index: int
+    ) -> T.Tuple[
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        T.Union[torch.Tensor, None]
+    ]:
+        pad_slice2d = (
+            slice(
+                int(batch.row_pad_before[batch_index]),
+                int(batch.height[batch_index])-int(batch.row_pad_after[batch_index])
+            ),
+            slice(
+                int(batch.col_pad_before[batch_index]),
+                int(batch.width[batch_index])-int(batch.col_pad_after[batch_index])
+            )
+        )
+        pad_slice3d = (
+            slice(0, None),
+            slice(
+                int(batch.row_pad_before[batch_index]),
+                int(batch.height[batch_index])-int(batch.row_pad_after[batch_index])
+            ),
+            slice(
+                int(batch.col_pad_before[batch_index]),
+                int(batch.width[batch_index])-int(batch.col_pad_after[batch_index])
+            )
+        )
+        rheight = pad_slice2d[0].stop - pad_slice2d[0].start
+        rwidth = pad_slice2d[1].stop - pad_slice2d[1].start
+        def reshaper(x: torch.Tensor, channel_dims: int) -> torch.Tensor:
```

Comment:
might be good to introduce an autoformatter like `black`

Reply:
good idea